### DISCUSSION TITLE
digital output for staubli tx60

### DISCUSSION THREAD

**Discussion by moutemio** [visose/Robots#285] [Q&A] [answered]
![IMG_20240708_155836](https://github.com/visose/Robots/assets/116132822/3e7cb34c-0a57-4052-b460-93b3d8d6603e)
Hi,

I am trying to control the digital output "0" of the CS8C controller for the TX60 robot. However, every time I attempt to run the VAL3 program, I encounter the following error message: "ERR 123 IO not linked to a hardware input-output."

I am uncertain about how to resolve this issue. I have already modified the XML file to match the actual digital output of the controller, but the error persists.

Could someone please guide me on how to correctly link the digital output to the hardware input-output using the plugins? Any help or suggestions would be greatly appreciated.

Thank you.

**Comment by visose** [comment]
Hi @moutemio have a look at this thread: https://github.com/visose/Robots/discussions/239
Could be related to the comments there about `<IO useControllerNumbering="true">`

**Reply by visose** [reply]
It looks like the poster deleted most of his messages for some reason. My answers are still there, but it's like hearing only one side of a conversation (the more useful side at least).

**Reply by moutemio** [reply]
Hi,

I've already seen that thread and, based on the comments there, I made some changes to the XML file (renaming the digital outputs to 1, 2, 3, 4, etc.). After that, another error appeared saying "ERR 123 IO not linked to a hardware input-output." I referred to the VAL3 manual of Staubli robots and managed to fix that issue. However, I am now encountering another error: "missing or invalid name or keyword."

I am currently stuck and would greatly appreciate any assistance you can provide. Attached is the PGX file I am using.

Thank you for your help.

Best regards,
Orel
![image](https://github.com/visose/Robots/assets/116132822/05fa1e54-3e02-4c70-8149-cac0375b9507)
![IMG_20240709_093609](https://github.com/visose/Robots/assets/116132822/842ec404-c042-4df7-84dc-ab26751102a9)

**Reply by visose** [reply]
Did you use a custom command? Not sure about the line `dioLink(..)`. Could you post a minimal Grasshopper file that reproduces the issue? (Use only 1 target, no external plugins besides Robots, no unecessary components).

**Reply by moutemio** [reply]
yes i used the custom command here is the file 
[visose.zip](https://github.com/user-attachments/files/16141045/visose.zip)

**Reply by visose** [reply]
Can you confirm if the error is due to your custom command?

**Reply by moutemio** [reply]
when i removed the custom command i return to the first error "io not linked to a hardware input-ouput" so it's not due to the custom command that's why i added the line "diolink (dos[1], io:valve1)" to link dos [1] to the output valve1 but i get another error  " 'io' : invalid library identifier"

**Reply by moutemio** [reply]
Could the problem came from the hardware configuration ?

**Reply by visose** [reply]
I meant if the last error you are getting is due to the custom command. You shouldn't need one.

**Comment by visose** [comment]
> Could you post a minimal Grasshopper file that reproduces the issue? (Use only 1 target, no external plugins besides Robots, no unecessary components).

![image](https://github.com/visose/Robots/assets/5129614/4e66fb6c-5b7a-4930-a90f-ac0b2642233c)
🤦🏻 

Now I'm wondering if you are accidentally mixing components from different robot plugins.

**Comment by moutemio** [comment, marked as answer]
as i investigated the solution i found out we have to link the DO directly on the teach pendant of the robot now it works !!!
